### PR TITLE
[FLINK-5214] Clean up checkpoint data in case of a failing checkpoint operation

### PR DESCRIPTION
**Pull request by tillrohrmann**
Adds exception handling to the stream operators for the snapshotState method. A failing
snapshot operation will trigger the clean up of all so far generated state resources.
This will avoid that in case of a failing snapshot operation resources (e.g. files) are
left behind.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContext.java [hunk 1]
```diff
@@ -21,8 +21,9 @@
 import org.apache.flink.annotation.PublicEvolving;
 
 /**
- * This interface provides a context in which operators that use managed (i.e. state that is managed by state
- * backends) or raw (i.e. the operator can write it's state streams) state can perform a snapshot.
+ * This interface provides a context in which operators that use managed (i.e. state that is managed
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContext.java]: All changes in this file are unrelated/reformatting.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java [hunk 2]
```diff
@@ -127,4 +128,38 @@ public OperatorStateCheckpointOutputStream getRawOperatorStateOutput() throws Ex
...
-}
\ No newline at end of file
+	private <T extends StreamStateHandle> void closeAndUnregisterStream(NonClosingCheckpointOutputStream<T> stream) throws IOException {
+		Preconditions.checkNotNull(stream);
+
+		closableRegistry.unregisterClosable(stream.getDelegate());
+		stream.getDelegate().close();
+	}
+
+	public void close() throws IOException {
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java]: If we already introduce a close method, we could also make this implement `AutoCloseable` and use it in try-with-resource?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 3]
```diff
@@ -342,21 +343,50 @@ public final OperatorSnapshotResult snapshotState(
 		StateSnapshotContextSynchronousImpl snapshotContext = new StateSnapshotContextSynchronousImpl(
 				checkpointId, timestamp, streamFactory, keyGroupRange, getContainingTask().getCancelables());
 
-		snapshotState(snapshotContext);
+		try {
```
Comment (StefanRRichter) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Here we could use the try-with-resource if `SnapshotContext` implements `AutoCloseable`

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContext.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContext.java]: True. Will revert them.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java]: Good idea. I will let it implement the `Closeable` interface

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Yes, you're right. Will refactor the code that way :-)

**Comment by tillrohrmann** [comment]
Thanks for the review @StefanRRichter. I'll address your comments and then merge this PR.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java (hunk 2 above)
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java]: I would prefer `AutoCloseable` over `Closeable`, because the former is general while the latter is more related to IO.

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java]: But isn't it better to be as specific as possible? In this case we wouldn't unnecessarily widen the thrown exception.

Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateSnapshotContextSynchronousImpl.java]: Of course, but you can and should always pick a more specific exception on the signature of your implementation. If `IOException` or a subclass is appropriate here, then you could also go with `Closeable` of course.
